File: dsa/concepts/13_graphs/java/UnionFind_L684_RedundantConnection.java

import java.util.Arrays;

public class UnionFind_L684_RedundantConnection {

	public static void main(String[] args) {
		UnionFind_L684_RedundantConnection obj = new UnionFind_L684_RedundantConnection();

		int[][] edges = {{1,2},{1,3},{2,3}};
		int[] redundant = obj.findRedundantConnection(edges);
		System.out.println(Arrays.toString(redundant));
	}

    // Problem: https://leetcode.com/problems/redundant-connection/
    // Idea: Duplicate edge creates a cycle in the graph. Use Union Find to find the cycle.
	public int[] findRedundantConnection(int[][] edges) {
        int n = edges.length+1;     // as problem said to have equal edges as nodes
        int[] parent = new int[n];
        int[] rank = new int[n];

        for(int i=1; i<n; i++){
            parent[i] = i;
            rank[i] = 0;
        }

        for(int[] edge: edges){
            int srcPar = find(parent, edge[0]);
            int dstPar = find(parent, edge[1]);
            if(srcPar == dstPar){
                return edge;
            }else{
                unionByRank(parent, rank, srcPar, dstPar);
            }
        }
        return new int[]{};
    }

    private int find(int[] parents, int curr){
        if(parents[curr] == curr)
            return parents[curr];
        return parents[curr] = find(parents, parents[curr]);
    }

    private void unionByRank(int[] parents, int[] ranks, int src, int dst){
        if(ranks[src] > ranks[dst]){
            parents[dst] = src;
        }else if(ranks[dst] > ranks[src]){
            parents[src] = dst;
        }else{
            parents[dst] = src;
            ranks[src] = ranks[src] + 1;
        }
    }

}

File: dsa/concepts/13_graphs/python/union_find_l684_redundant_connection.py
from typing import List


class UnionFind:
    def __init__(self, n: int):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x: int, y: int) -> bool:
        px, py = self.find(x), self.find(y)
        if px == py:
            return False

        if self.rank[px] < self.rank[py]:
            self.parent[px] = py
        elif self.rank[px] > self.rank[py]:
            self.parent[py] = px
        else:
            self.parent[py] = px
            self.rank[px] += 1

        return True


"""
Problem:
Redundant Connection

Idea:
- The graph initially forms a tree.
- Adding one extra edge introduces a cycle.
- While processing edges, if two nodes already belong
  to the same set, that edge is redundant.

Approach:
- Use Union-Find with path compression.
- For each edge:
    - If both nodes have same parent → redundant edge
    - Else union them

Time:
O(n α(n))

Space:
O(n)
"""
def find_redundant_connection(edges: List[List[int]]) -> List[int]:
    n = len(edges) + 1
    uf = UnionFind(n)

    for u, v in edges:
        if not uf.union(u, v):
            return [u, v]

    return []


# ------------ test (same as Java main) ------------
if __name__ == "__main__":
    edges = [[1, 2], [1, 3], [2, 3]]
    print(find_redundant_connection(edges))
